server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -78,6 +79,7 @@ public class AllocationService {
     private final ClusterInfoService clusterInfoService;
     private final SnapshotsInfoService snapshotsInfoService;
     private final ShardRoutingRoleStrategy shardRoutingRoleStrategy;
+    private final ThreadPool threadPool;
```

Comment:
This is a pretty heavy dependency - setting up and tearing down a whole `ThreadPool` for each test is too expensive for my liking. Could we pass in a `LongSupplier` instead, injecting `threadPool::relativeTimeInNanos` only where necessary? The only unit tests that care about time should be the ones that use a `MockAllocationService` for which we can pass in `() -> nanoTimeOverride`, and the others can just use `() -> 0L`.

Reply:
Re `LongSupplier`. I wonder if it is better to introduce dedicated interface for this, similar to java.time.InstantSource or java.time.Clock with dedicated methods to retrieve time? (relativeTimeInMillis/relativeTimeInNanos/rawRelativeTimeInMillis/absoluteTimeInMillis).

Having dedicated interface would:
* make it clear what kind of time is actually used
* make it easier to have a simple test implementation

Reply:
Sure, we could do that too. I'm not convinced we want `absoluteTimeInMillis` in there - the places that care about absolute time are rather different from the ones that care about relative time. Also not sure how big a change that would be, we use `LongSupplier` quite a bit already today.

Reply:
> Also not sure how big a change that would be

May be it is worth a separate pr with general time injection improvements if that is too big.

> LongSupplier quite a bit already today.

I am not sure this is good in general as does not make it obvious if that is millis or nanos nor if it is cached or raw.
However I do not mind using that if above turns out to be a big change on its own.

Reply:
Thank you for your input! I changed thread pool in favor of LongSupplier. In ClusterModule it is bound to ThreadPool#relativeTimeInNanos whereas in tests - to System::nanoTime.

Regarding number of affected files. 
I deliberately did not introduce a new CTOR for AllocationService in order to set default value (read it 'hide') for LongSupplier param and hence leave existing AllocationService's referees (see them in tests) untouched. I think that newly introduced LongSupplier is important part of code contract and it should stay there until resolution of #95800